Sources/Bond/Data Structures/TreeProtocol.swift

```diff
@@ -0,0 +1,178 @@
+//
+//  The MIT License (MIT)
+//
+//  Copyright (c) 2018 DeclarativeHub/Bond
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+/// A protocol that provides abstraction over a tree type.
+/// A tree can be any containter type that encapsulates objects or values that are also trees.
+public protocol TreeProtocol {
+
+    /// A collection of child nodes that are trees and whose children are also trees
+    associatedtype Children: Collection where
+        Children.Element: TreeProtocol,
+        Children.Element.Children == Children,
+        Children.Index == Int
+
+    /// Child nodes of the current tree node.
+    var children: Children { get }
+}
+
+public protocol MutableTreeProtocol: TreeProtocol where Children: MutableCollection, Children.Element: MutableTreeProtocol {
+
+    /// Child nodes of the current tree node.
+    var children: Children { get set }
+}
+
+public protocol RangeReplaceableTreeProtocol: MutableTreeProtocol {
```

Comment:
Why is this reliant on the actual `children` property being replaceable? Surely it's only the contents that needs to be?